paws.common/tests/testthat/test_credentials.R

```diff
@@ -14,24 +14,130 @@ test_that("credentials not provided", {
 
   expect_false(is_credentials_provided(creds))
 
-  creds <- list(
+  creds <- Creds(
     access_key_id = "",
     secret_access_key = "bar"
   )
 
   expect_false(is_credentials_provided(creds))
 
-  creds <- list(
+  creds <- Creds(
     access_key_id = "foo",
     secret_access_key = ""
   )
 
   expect_false(is_credentials_provided(creds))
 
-  creds <- list(
+  creds <- Creds(
     access_key_id = "",
     secret_access_key = ""
   )
 
   expect_false(is_credentials_provided(creds))
 })
+
+test_that("credentials expired", {
+  creds <- Creds(
+    access_key_id = "foo",
+    secret_access_key = "bar",
+    expiration = 0
+  )
+
+  expect_false(is_credentials_provided(creds))
+
+  creds <- Creds(
+    access_key_id = "foo",
+    secret_access_key = "bar",
+    expiration = Inf
+  )
+
+  expect_true(is_credentials_provided(creds))
+
+  creds <- Creds(
+    access_key_id = "foo",
+    secret_access_key = "bar",
+    expiration = NA
+  )
+
+  expect_true(is_credentials_provided(creds))
+
+})
+
+#-------------------------------------------------------------------------------
+
+# Example operation that gets credentials (if unavailable or expired) and
+# returns the credentials.
+f_get_credentials <- function() {
+  cfg <- get_config()
+  if (!is_credentials_provided(cfg$credentials$creds, window = 0)) {
+    foo <- get_credentials(cfg$credentials)
```

Comment:
This test works but I am a little confused about what is happening since `foo` isn't being used. Is it because `get_credentials` sets the value and `cfg$credentials` is just a reference to that environment so it will still be pointed at the correct value?

Reply:
That's exactly right. I'll add comments that it sets the credentials as a side effect, and get rid of the assignment to `foo`.